### PR TITLE
Feature d200 lidar

### PR DESCRIPTION
**Pull request by saaaji**
basic d200 lidar driver is up and running. works in arduino IDE but haven't been able to see if it builds with the rest of the codebase.

### CODE REVIEW

File: src/sensors/d200.hpp
Comment (StudentKyle) [inline src/sensors/d200.hpp:125]: What happens if there is not new data to be read? Does it give duplicate packets (ideally labeled with the old teensy timestamp)? Or something else? Please document this behavior as well as make sure it's correct.

File: src/sensors/d200.hpp
Comment (StudentKyle) [inline src/sensors/d200.hpp:82]: Can you instead just throw out this data? This should be assigned to the teensy time instead (through micros() or millis(), but always converted into seconds for the final return)

File: src/sensors/d200.hpp
Comment (StudentKyle) [inline src/sensors/d200.hpp:68]: Convert all units of degrees into radians, all units of time into seconds

File: src/sensors/d200.hpp
Comment (StudentKyle) [inline src/sensors/d200.hpp:70]: convert to radians

File: src/sensors/d200.hpp
Comment (StudentKyle) [inline src/sensors/d200.hpp:75]: convert distances into meters

File: src/sensors/d200.hpp
Comment (StudentKyle) [inline src/sensors/d200.hpp:79]: convert to radians

File: src/sensors/d200.hpp
Comment (StudentKyle) [inline src/sensors/d200.hpp:77]: Normalized into what? What is the expected range? Does a low value represent a low or high intensity?

File: src/sensors/d200.hpp
Comment (StudentKyle) [inline src/sensors/d200.hpp:125]: Make sure the end result is converted into radians, meters, and seconds, since that is what the rest of the project expects from every sensor.

You could have LidarDataPacket be the raw data as you currently have parsed, and then return a different struct with the filtered data into converted units. Could also be a good post-process step to throw out sensor time and give the teensy time instead (in seconds). 

**Addendum:** Make sure the timestamp is set when the complete packet is read, NOT when gotten in the getter here.

File: src/sensors/d200.hpp
Comment (StudentKyle) [inline src/sensors/d200.hpp:110]:
```suggestion
    /// @brief set the rotation speed of the LiDAR
```

File: src/sensors/d200.cpp
Comment (StudentKyle) [inline src/sensors/d200.cpp:62]: Throw out timestamp from lidar, instead use teensy time. Can be done here or in a separate post-process step. I'd prefer a separate post-process step that happens as soon as the read() function reads a whole packet, that can also do things like unit conversions into a result struct.

File: src/sensors/d200.cpp
Comment (StudentKyle) [inline src/sensors/d200.cpp:32]: Does this start/stop it from spinning?

File: src/sensors/d200.cpp
Comment (StudentKyle) [inline src/sensors/d200.cpp:35]: Does this start/stop it from spinning?

File: src/sensors/d200.cpp
Comment (StudentKyle) [inline src/sensors/d200.cpp:40]: I really like how you handled this to support any length of serial buffer. However, two edge cases to fix/verify:

**Case 1:**
1. You have a partially filled buffer, with say, 20 bytes in it, from a previous read() call, and are awaiting the next 27 bytes.
2. Time passes such that LIDAR_SERIAL gets full and overwrites old data multiple times. 
    Now, the partially filled packet would be invalid to continue filling, but code doesn't know that yet.
3. Read() is called some time later. Maybe a new start byte is encountered, maybe not, but regardless your code currently will not reset the buffer index = 0. Instead, it will read the next 27 garbage bytes as valid data. The checksum might catch it, but this isn't guarenteed since measurement data section could be anything (unless I am misunderstanding the checksum standard).
This edge case looks broken to me. 

I think this could be fixed by checking if LIDAR_SERIAL.available() == MAX_SERIAL_BUFFER. If true, it is possible something was overwritten. So, throw out the in-progress packet (reset idx = 0) and begin checking for a new start character instead.
This solution comes at a cost of lost packets, but I think correctness is more important here since this shouldn't happen too often unless the main loop code is too slow.


**Case 2:**
LIDAR_SERIAL has more than PACKET_SIZE bytes, so two possibilities exist:
there is one complete packet and one incomplete packet, or two incomplete packets. 

I ran through possibilities in my head and it looks like your solution functions, but you should make sure this performs as expected (don't repeatedly drop packets, etc).

File: src/sensors/d200.hpp
Comment (StudentKyle) [inline src/sensors/d200.hpp:96]: Consider adding a function that converts a LidarPacket into a point cloud (list of x,y(,z) points), using angle + distance transformation. Perhaps this isn't needed on the teensy side, but it could be useful on the Khadas.

**Review by StudentKyle — CHANGES_REQUESTED** [review]
Reviewed the code. Make sure to convert to the correct units, use the teensy timestep, and handle the edge case I described in one of the comments. There's also some smaller code quality things to fix.
 
Other than that I think you came up with a very clean solution to handle the lidar reading!

**Review by nitroxplunge — CHANGES_REQUESTED** [review]
Improvements can be made when reading data

File: src/sensors/d200.cpp
Comment (nitroxplunge) [reply src/sensors/d200.cpp:40]: I disagree with this.

File: src/sensors/d200.cpp
Comment (nitroxplunge) [reply src/sensors/d200.cpp:56]: Seems inherit to the data structure in which case the magic numbers are fine, but they require a comment for sure

File: src/sensors/d200.cpp
Comment (nitroxplunge) [reply src/sensors/d200.cpp:62]: Nope nope nope. Use the native time of the sensor. Using teensy time will result in synchronization issues

File: src/sensors/d200.hpp
Comment (nitroxplunge) [reply src/sensors/d200.hpp:46]: This is fine

File: src/sensors/d200.cpp
Comment (StudentKyle) [reply src/sensors/d200.cpp:62]: @nitroxplunge The sensor time wraps around at 30s, making it basically useless on the scale of a match. Most simple is to switch to teensy time, and we don't need to handle what happens when we go from 29.9s to 0.1s. 

Correct me if I'm wrong, but "lidar time" doesn't give us any more useful information than teensy time does. And it has the extra limitation of wrapping at 30s, so we just use the better option instead.

File: src/sensors/d200.cpp
Comment (StudentKyle) [reply src/sensors/d200.cpp:40]: You disagree with the edge case existing?

**Comment by StudentKyle** [comment]
Note you also have a couple undocumented members (state)

File: src/sensors/d200.hpp
Comment (StudentKyle) [reply src/sensors/d200.hpp]: resolved

File: src/sensors/d200.hpp
Comment (StudentKyle) [reply src/sensors/d200.hpp:139]: latest packet saved

File: src/sensors/d200.hpp
Comment (StudentKyle) [reply src/sensors/d200.hpp:101]: not necessary for lidar driver, but processing will need to happen before being able to use data.

File: src/sensors/d200.cpp
Comment (StudentKyle) [reply src/sensors/d200.cpp]: resolved

File: src/sensors/d200.cpp
Comment (StudentKyle) [reply src/sensors/d200.cpp]: Still need to figure this out.

File: src/sensors/d200.cpp
Comment (StudentKyle) [reply src/sensors/d200.cpp]: resolved

File: src/sensors/d200.cpp
Comment (StudentKyle) [reply src/sensors/d200.cpp]: discussed in meeting

File: src/sensors/d200.hpp
Comment (StudentKyle) [reply src/sensors/d200.hpp]: Still need to know. Making new comment on non-outdated version.

File: src/sensors/d200.hpp
Comment (StudentKyle) [reply src/sensors/d200.hpp]: Still need to figure out if "lidar time" from the packet actually gives useful information or not, or if teensy time should be used instead.

File: src/sensors/d200.hpp
Comment (StudentKyle) [reply src/sensors/d200.hpp]: resolved

File: src/sensors/d200.hpp
Comment (StudentKyle) [reply src/sensors/d200.hpp]: resolved

File: src/sensors/d200.hpp
Comment (StudentKyle) [reply src/sensors/d200.hpp]: resolved

File: src/sensors/d200.hpp
Comment (StudentKyle) [inline src/sensors/d200.hpp:89]: I read the docs. It says: 

"the higher the intensity, the larger the signal strength value; the lower the intensity, the smaller the signal strength value." Can this be added?

```suggestion
    /// @brief intensity of measurement. The higher the intensity, the larger the signal strength value.
```

Comment (StudentKyle) [reply src/sensors/d200.hpp:89]: It's likely it's just bound by the 1-byte limit unless you found otherwise via testing.

File: src/sensors/d200.hpp
Comment (StudentKyle) [inline src/sensors/d200.hpp:97]: We still need to figure out what to do with the timestamp. Since it wraps at 30s, the value provided from the lidar would not be useful over the course of a match without additional processing (increase wrap range from this provided time, or just use teensy time).

For WorldMap I need a timestamp that is reliable over the entire course of a match, and be one that I can compare with a current time (either lidar or teensy current time). However you think that can happen is fine, but if we use teensy time we won't need to send an additional timestamp representing "current time" just for lidar through comms. 

(Regardless we'll need to send some kind of lidar timestamp through so the points have some time associated with them)

File: src/sensors/d200.hpp
Comment (StudentKyle) [inline src/sensors/d200.hpp:151]: What happens if there are no packets ready? Does it return the "previous" latest packet? This behavior should be documented. 

(Currently it looks like it'll crash if the array is empty. Could the array be empty during normal execution (if packet array is cleared)? What about at the very start of execution?).

File: src/sensors/d200.hpp
Comment (StudentKyle) [reply src/sensors/d200.hpp]: See new comment.

File: src/sensors/d200.cpp
Comment (StudentKyle) [reply src/sensors/d200.cpp]: resolved

File: src/sensors/d200.cpp
Comment (StudentKyle) [reply src/sensors/d200.cpp]: resolved.

File: src/sensors/d200.cpp
Comment (StudentKyle) [reply src/sensors/d200.cpp]: Code was restructured, this analysis no longer applies

**Review by StudentKyle — CHANGES_REQUESTED** [review]
Updated standing issues from previous review.

File: src/sensors/d200.hpp
Comment (StudentKyle) [inline src/sensors/d200.hpp:78]: Undocumented type

**Review by StudentKyle — CHANGES_REQUESTED** [review]
Undocumented type

File: src/sensors/d200.hpp
Comment (siraajsandhu) [reply src/sensors/d200.hpp:151]: won't crash, the packets are zero initialized but I agree the behavior can be better documented.

Comment (saaaji) [reply src/sensors/d200.hpp:151]: won't crash, the packets are zero initialized but I agree the behavior can be better documented.

File: src/sensors/d200.hpp
Comment (StudentKyle) [reply src/sensors/d200.hpp:97]: Another option is to say timestamps wrapping is just fine, and the WorldMap will then handle the wrap on its level:
```python
if (new_lidar_timestamp < prev_lidar_timestamp) lidar_time =  new_lidar_timestamp + LIDAR_TIMESTAMP_WRAP
```

I don't think that kind of behavior belongs in WorldMap. It may be cleaner if all the sources of data (realsense, robot_state, lidar) just have their own, accurate timestamp handling the wrap themselves.

Comment (StudentKyle) [reply src/sensors/d200.hpp:97]: What about a khadas-level lidar interface, that could perform the timestamp unwrapping as well as constructing the structured Point Cloud? I'd be fine with that and then it doesn't have to be a part of the lidar driver in this PR.

**Review by Rishi-H — APPROVED** [review]

**Review by nitroxplunge — APPROVED** [review]

**Review by StudentKyle — CHANGES_REQUESTED** [review]
You have undocumented members. Please see doxygen warning output

**Comment by StudentKyle** [comment]
Overall it looks good, but I can't comment on the functionality of the packet interpretation in particular. If it works and the latest packet is accessible as a LidarPacketData then good, but this should be tested.

**Comment by StudentKyle** [comment]
I am still not convinced we want to be taking the lidar-packet-provided timestamp over the teensy's time, since if the lidar ever stops, we won't be able to correctly compare age on the khadas side.

Please make this decision (whichever you end up providing in the return) clear in the documentation: are we getting a lidar packet time that wraps at 30s, or is the timestamp the same one as the rest of the teensy?

**Comment by StudentKyle** [comment]
After speaking with @Pandabear1125, hive can operate on the assumption the lidar won't stop. If this assumption fails, this can be a problem for later since fixing it might cost Comms bandwidth. 

I'll approve my review after the docs are fixed, and defer the testing to another reviewier. Also, Jackson should review the export_data for Comms, since 8*45 bytes is likely too much for Comms.

File: src/sensors/d200.hpp
Comment (StudentKyle) [inline src/sensors/d200.hpp:175]:
```suggestion
    /// @brief export LiDAR data as byte array for comms. Exports the latest D200_NUM_PACKETS_CACHED packets, for a total size of D200_NUM_PACKETS_CACHED * D200_PAYLOAD_SIZE bytes per export.
```

Comment (StudentKyle) [reply src/sensors/d200.hpp:175]: clarify how large a byte array is given in documentation.

**Review by StudentKyle — CHANGES_REQUESTED** [review]

**Review by StudentKyle — APPROVED** [review]
lgtm. We can hash out any problems via testing on the robot

**Pull request merged**
